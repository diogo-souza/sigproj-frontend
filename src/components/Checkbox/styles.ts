import styled, { css } from 'styled-components';
import { CheckboxProps } from '.';

type ContainerProps = Pick<CheckboxProps, 'labelColor' | 'fillColor'>;

export const Label = styled.label<ContainerProps>`
  ${({ theme, labelColor }) => css`
    cursor: pointer;
    display: flex;
    align-items: center;
    > span {
      padding-left: ${theme.spacings.xxsmall};
      color: ${theme.colors[labelColor!]};
      line-height: 1.8rem;
    }
  `}
`;

export const Input = styled.input<ContainerProps>`
  ${({ theme, fillColor, labelColor }) => css`
    cursor: pointer;
    display: flex;
    align-items: center;
    justify-content: center;
    appearance: none;
    width: 1.8rem;
    height: 1.8rem;
    border: 0.2rem solid ${theme.colors[labelColor!]};
    border-radius: 0.2rem;
    transition: background border 0.1s ease-in-out;
    position: relative;
    outline: none;
    &:before {
      content: '';
      width: 0.6rem;
      height: 0.9rem;
      border: 0.2rem solid ${theme.colors.white};
      border-top: 0;
      border-left: 0;
      transform: rotate(45deg);
      position: absolute;
      top: 0.1rem;
      opacity: 0;
      transition: 0.1s ease-in-out;
    }
    &:focus {
      box-shadow: 0 0 0.5rem ${theme.colors[fillColor!]};
    }
    &:hover {
      border-color: ${theme.colors.gray};
      transition: 0.1s ease-in-out;
    }
    &:checked {
      border-color: ${theme.colors[fillColor!]};
      background: ${theme.colors[fillColor!]};
      &:before {
        opacity: 1;
      }
    }
  `}
`;
