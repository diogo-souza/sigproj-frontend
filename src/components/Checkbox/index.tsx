import React, { InputHTMLAttributes, useState } from 'react';
import * as S from './styles';

export type CheckboxProps = {
  onCheck?: (status: boolean) => void;
  isChecked?: boolean;
  label: string;
  labelFor: string;
  labelColor?: 'titleBlack' | 'titleWhite';
  fillColor?: 'primary' | 'secondary' | 'tertiary';
  value?: string | ReadonlyArray<string> | number;
} & InputHTMLAttributes<HTMLInputElement>;

const Checkbox: React.FC<CheckboxProps> = ({
  onCheck,
  isChecked = false,
  label,
  labelFor,
  labelColor = 'titleBlack',
  value,
  fillColor = 'secondary',
  ...props
}: CheckboxProps) => {
  const [checked, setChecked] = useState(isChecked);

  const onChange = () => {
    const status = !checked;
    setChecked(status);

    // eslint-disable-next-line no-unused-expressions
    !!onCheck && onCheck(status);
  };

  return (
    <S.Label htmlFor={labelFor} labelColor={labelColor}>
      <S.Input
        id={labelFor}
        onChange={onChange}
        type="checkbox"
        checked={checked}
        value={value}
        fillColor={fillColor}
        {...props}
      />
      <span>{label}</span>
    </S.Label>
  );
};

export default Checkbox;
