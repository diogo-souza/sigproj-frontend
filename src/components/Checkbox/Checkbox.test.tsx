import { screen, waitFor } from '@testing-library/react';
import { renderWithTheme } from 'utils/tests/helpers';
import theme from 'styles/themes';

import userEvent from '@testing-library/user-event';
import Checkbox from '.';

describe('<Checkbox />', () => {
  it('should render with label', () => {
    const { container } = renderWithTheme(
      <Checkbox label="checkbox label" labelFor="check" />,
    );
    expect(screen.getByRole('checkbox')).toBeInTheDocument();
    expect(screen.getByLabelText(/checkbox label/i)).toBeInTheDocument();
    expect(
      screen.getByText(/checkbox label/i).closest('label'),
    ).toHaveAttribute('for', 'check');

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with black label', () => {
    renderWithTheme(
      <Checkbox
        label="checkbox label"
        labelFor="check"
        labelColor="titleBlack"
      />,
    );
    expect(screen.getByText(/checkbox label/i)).toHaveStyle({
      color: theme.colors.titleBlack,
    });
  });

  it('should render with white label', () => {
    renderWithTheme(
      <Checkbox
        label="checkbox label"
        labelFor="check"
        labelColor="titleWhite"
      />,
    );
    expect(screen.getByText(/checkbox label/i)).toHaveStyle({
      color: theme.colors.titleWhite,
    });
  });

  it('should render with primary fillColor', () => {
    renderWithTheme(
      <Checkbox
        label="checkbox label"
        labelFor="check"
        labelColor="titleWhite"
        fillColor="primary"
        isChecked
      />,
    );
    expect(screen.getByRole('checkbox')).toHaveStyle({
      background: theme.colors.primary,
    });
  });

  it('should render with secondary fillColor', () => {
    renderWithTheme(
      <Checkbox
        label="checkbox label"
        labelFor="check"
        labelColor="titleWhite"
        fillColor="secondary"
        isChecked
      />,
    );
    expect(screen.getByRole('checkbox')).toHaveStyle({
      background: theme.colors.secondary,
    });
  });

  it('should render with tertiary fillColor', () => {
    renderWithTheme(
      <Checkbox
        label="checkbox label"
        labelFor="check"
        labelColor="titleWhite"
        fillColor="tertiary"
        isChecked
      />,
    );
    expect(screen.getByRole('checkbox')).toHaveStyle({
      background: theme.colors.tertiary,
    });
  });

  it('should dispatch onCheck when status change', async () => {
    const onCheck = jest.fn();
    renderWithTheme(
      <Checkbox label="checkbox label" labelFor="check" onCheck={onCheck} />,
    );

    expect(onCheck).not.toHaveBeenCalled();

    userEvent.click(screen.getByRole('checkbox'));
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });

    expect(onCheck).toHaveBeenCalledWith(true);
  });

  it('should call onCheck with false if the Checkbox is already checked', async () => {
    const onCheck = jest.fn();
    renderWithTheme(
      <Checkbox
        label="checkbox label"
        labelFor="check"
        onCheck={onCheck}
        isChecked
      />,
    );

    expect(onCheck).not.toHaveBeenCalled();

    userEvent.click(screen.getByRole('checkbox'));
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });

    expect(onCheck).toHaveBeenCalledWith(false);
  });

  it('should be accessible with tab', async () => {
    renderWithTheme(<Checkbox label="checkbox label" labelFor="check" />);

    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(screen.getByLabelText(/checkbox label/i)).toHaveFocus();
  });
});
