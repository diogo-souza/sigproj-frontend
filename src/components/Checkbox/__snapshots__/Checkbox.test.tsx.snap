// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render with label 1`] = `
.c0 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > span {
  padding-left: 0.8rem;
  color: #171923;
  line-height: 1.8rem;
}

.c1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 1.8rem;
  height: 1.8rem;
  border: 0.2rem solid #171923;
  border-radius: 0.2rem;
  -webkit-transition: background border 0.1s ease-in-out;
  transition: background border 0.1s ease-in-out;
  position: relative;
  outline: none;
}

.c1:before {
  content: '';
  width: 0.6rem;
  height: 0.9rem;
  border: 0.2rem solid #FFFFFF;
  border-top: 0;
  border-left: 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
  position: absolute;
  top: 0.1rem;
  opacity: 0;
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
}

.c1:focus {
  box-shadow: 0 0 0.5rem #159CAE;
}

.c1:hover {
  border-color: #939393;
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
}

.c1:checked {
  border-color: #159CAE;
  background: #159CAE;
}

.c1:checked:before {
  opacity: 1;
}

<label
  class="c0"
  for="check"
>
  <input
    class="c1"
    id="check"
    type="checkbox"
    value=""
  />
  <span>
    checkbox label
  </span>
</label>
`;
